tutorials/6_comp_invars.md

```diff
@@ -317,23 +317,26 @@ constructor. In reality, all other representatives are zero initialized, and may
 therefore be in a different state than those in the neighbourhood. We must check
 that all such neighbourhoods satisfy the invariant.
 
-Finally, `Non-interference` check generalizes the challenge faced by the
-`Initialization` check. Here we must consider two neighbourhoods. One
-neighbourhood is fixed from before the transition, while the other takes part
-directly in the transition. However, simply fixing an external neighbourhood is
-not enough. If we fix the neighbourhood, we miss all neighbourhoods for which
-the acting process and the external process share one or more client. In effect,
-we must "decide" which clients the two neighbourhoods share, and then update the
-state as required.
-
-We address the forementioned challenge with two modeling tricks.
+The `Non-interference` check generalizes the challenge faced by the
+`Initialization` check. Here we must consider pairs of neighbourhoods. One
+neighbourhood is fixed before the transition takes place, while the other takes
```

Comment:
Is it just one neighbourhood before? Or should it be 0 or more? This models the "interference", right?

Reply:
We assert the invariant before the transition. Why can't we assume the invariant instead?

Reply:
If there are n neighbourhoods and you want to consider between 0 and n "interference steps" (up to 1 per neighbourhood) this gives you the exponential blowup in the naive encoding, right?

Reply:
To the first point: We have two neighbourhoods, A and B, each with n representative clients. We are allowing A to make a single move. We want to show that the compositional invariant of B still holds after A has made a move. If A and B share some client c, then updates made by A to client c should be reflected in B after A has made its move. This is the overlap I'm talking about. Since there are n reps., there are 2^n possible sets of overlaps. -- The overlap is used to model one part of the interference. Without the overlap, we still have interference (the global state of the contract can change).  -- Honestly, I wasn't entirely clear on this question but I hope this clears it up. If this does clear it up, please let me know which points were unclear. I will reflect those in the document.

To the second point: I did not mean to suggest that we make assertions before the transition. Do you might highlighting which part of the text suggested otherwise to you? I can try to clarify that part.

To the third point: This is a single step. We are looking at interference between neighbourhoods rather than interference between single clients. I think this might have been part of the misunderstanding in the first point.